File: packages/core/src/conformance/history-center.ts
/**
 * HistoryCenter portable interface — composite profile (g14.007).
 *
 * Contract: `docs/contracts/components/history-center.md`. This is the first
 * host-coordinated composite in the conformance kernel: the component owns no
 * history, validates no protocol rule, and decides nothing about undo. Data
 * arrives as `pages`; every operation leaves as a named command and comes back
 * as host-supplied records.
 *
 * Three things make that expressible without a second conformance
 * architecture, and all three are generic kernel vocabulary:
 *
 * - **Structured fixture data.** `pages` is a collection whose `entries` field
 *   is itself a collection, so the page → entry shape the authority hands over
 *   survives into the fixture instead of being flattened into something the
 *   component never sees. One nesting level; deeper is an authoring error.
 * - **Host record channels.** `continuations` and `runEntries` are the data a
 *   fixture host answers `loadContinuations` and `loadContinuationRun` from.
 *   They are declared — so their identity is validated and repeated parts key
 *   over them — but they are not portable props: the component receives them
 *   through its callbacks' results, never as inputs, so they generate no Rust
 *   spec field.
 * - **Repeat sources.** A row is a row whether it came from the spine the host
 *   passed in or from a run the host answered with, so `row`, `entry` and
 *   `disclosure` key over both sources into one identity space. A fork run
 *   entry that repeated a spine entry id is rejected at authoring — that
 *   collision is exactly the duplicate-row defect the v3 derivation exists to
 *   prevent.
 *
 * Longhorn is absent and cannot arrive: the dependency runs Longhorn → Poodle,
 * the record shapes here are structural mirrors, and `checkout` is Poodle's own
 * word for what a host maps onto its own prefer operation.
 */

import { defineComponentInterface } from "./define";
import type {
  AxisNamesOf,
  EventNamesOf,
  PartIdsOf,
  PortableEventsOf,
  PortablePropsOf,
  TokenRoleNamesOf,
} from "./define";

/** One history entry, as the authority's `ForkEntryRecord` mirrors it.
 * `position` and the timestamps are host facts; nothing here invents a clock. */
const HISTORY_ENTRY_FIELDS = [
  { name: "id", type: { kind: "string" } },
  { name: "label", type: { kind: "string" } },
  // "past" | "current" | "future". Collection item fields carry no enum
  // vocabulary; the renderer projects the value onto `data-position` and the
  // corpus asserts the projection, so a wrong value is still observable.
  { name: "position", type: { kind: "string" } },
  { name: "checkpoint", type: { kind: "boolean" }, optional: true },
  { name: "groupId", type: { kind: "string" }, optional: true },
  { name: "recordedAtMs", type: { kind: "number" }, optional: true },
  // Every continuation from this entry, the run's own next row included. A
  // fork count is one less (R4); a run's last entry is always 0.
  { name: "continuationCount", type: { kind: "number" } },
] as const;

export const historyCenterInterface = defineComponentInterface({
  id: "history-center",
  profile: "composite",
  props: [
    {
      // Root path pages in fetch order, newest page first. `null` disables the
      // list: no rows render and every row event is inert.
      name: "pages",
      type: {
        kind: "collection",
        rustType: "HistoryPathPage",
        fields: [
          { name: "offset", type: { kind: "number" } },
          { name: "precedingContinuationCount", type: { kind: "number" } },
          { name: "truncatedBefore", type: { kind: "boolean" } },
          { name: "truncatedAfter", type: { kind: "boolean" } },
          {
            name: "entries",
            type: {
              kind: "collection",
              rustType: "HistoryEntry",
              fields: HISTORY_ENTRY_FIELDS,
            },
          },
        ],
      },
      default: null,
      nullable: true,
    },
    { name: "canUndo", type: { kind: "boolean" }, default: false, rustName: "can_undo" },
    { name: "canRedo", type: { kind: "boolean" }, default: false, rustName: "can_redo" },
    { name: "busy", type: { kind: "boolean" }, default: false, rustName: "is_busy" },
    {
      name: "status",
      type: { kind: "enum", values: ["idle", "loading", "failed"] },
      default: "idle",
      rustEnumName: "HistoryCenterStatus",
    },
    { name: "statusMessage", type: { kind: "string" }, default: null, nullable: true },
    {
      // A rejection code the host's bridge mapped from its protocol. The
      // component owns the display copy; the protocol never reaches Poodle.
      name: "rejection",
      type: { kind: "enum", values: ["AlreadyAtTarget", "UnknownEntry"] },
      default: null,
      nullable: true,
      rustEnumName: "HistoryCenterRejection",
    },
    {
      name: "open",
      type: { kind: "boolean" },
      default: null,
      nullable: true,
      controlledBy: "openChange",
    },
    { name: "defaultOpen", type: { kind: "boolean" }, default: false },
    {
      name: "placement",
      type: {
        kind: "enum",
        values: [
          "top", "top-start", "top-end",
          "bottom", "bottom-start", "bottom-end",
          "left", "left-start", "left-end",
          "right", "right-start", "right-end",
        ],
      },
      default: "bottom-end",
      rustType: "OverlayPlacement",
    },
    { name: "undoLabel", type: { kind: "string" }, default: "Undo" },
    { name: "redoLabel", type: { kind: "string" }, default: "Redo" },
    { name: "listLabel", type: { kind: "string" }, default: "History" },
    { name: "title", type: { kind: "string" }, default: "History" },
    { name: "emptyMessage", type: { kind: "string" }, default: "No history entries yet." },
    { name: "ariaLabel", type: { kind: "string" }, default: null, nullable: true },
    {
      // Client-side affordance only: it caps the rename input's length and
      // enforces no protocol rule.
      name: "maxBranchNameBytes",
      type: { kind: "number" },
      default: 256,
      rustType: "usize",
    },
    {
      name: "size",
      type: { kind: "enum", values: ["xs", "sm", "md", "lg", "xl"] },
      default: null,
      nullable: true,
      rustType: "ControlSize",
    },
    {
      name: "sizeRole",
      type: { kind: "enum", values: ["chrome", "control", "prominent"] },
      default: "chrome",
      rustType: "SemanticControlSizeRole",
    },
    {
      name: "density",
      type: { kind: "enum", values: ["compact", "default", "comfortable"] },
      default: null,
      nullable: true,
      rustType: "ControlDensity",
    },
  ],
  events: [
    { name: "undo", payload: {} },
    { name: "redo", payload: {} },
    { name: "openChange", payload: { open: "boolean" } },
    // Always the entry actually activated, on the branch that owns its run.
    // `branchId` is empty on the spine — the host knows its own branch.
    { name: "navigateEntry", payload: { branchId: "string", entryId: "string" } },
    { name: "renameBranch", payload: { branchId: "string", name: "string" } },
    { name: "loadContinuations", payload: { entryId: "string" } },
    { name: "loadContinuationRun", payload: { fromEntryId: "string" } },
    // Poodle's own word. A host maps it onto its own prefer operation.
    { name: "checkoutContinuation", payload: { entryId: "string" } },
  ],
  regions: [],
  hostRecords: [
    {
      // The forks at an anchor. The host returns every child of the anchor,
      // the one already on the list included; the derivation filters that one
      // out by id and never assumes its position.
      name: "continuations",
      answers: "loadContinuations",
      fields: [
        { name: "anchorEntryId", type: { kind: "string" } },
        { name: "entryId", type: { kind: "string" } },
        { name: "label", type: { kind: "string" } },
        { name: "preferred", type: { kind: "boolean" } },
        { name: "entryCount", type: { kind: "number" } },
        { name: "branchId", type: { kind: "string" } },
        { name: "branchName", type: { kind: "string" }, optional: true },
      ],
    },
    {
      // A continuation run's entries, newest first, tagged with the fork they
      // belong to. Flat rather than paged: the page-join reversal is proved on
      // the spine, and a run's own paging adds no new claim.
      name: "runEntries",
      answers: "loadContinuationRun",
      fields: [
        { name: "fromEntryId", type: { kind: "string" } },
        ...HISTORY_ENTRY_FIELDS,
      ],
    },
  ],
  parts: [
    { id: "root", resolve: { web: { kind: "self" }, native: { kind: "self" } } },
    {
      id: "undo",
      role: "button",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__undo .poodle-icon-button" },
        native: { kind: "id", id: "history-center:undo" },
      },
    },
    {
      id: "redo",
      role: "button",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__redo .poodle-icon-button" },
        native: { kind: "id", id: "history-center:redo" },
      },
    },
    {
      id: "list-trigger",
      role: "button",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__list-trigger" },
        native: { kind: "id", id: "history-center:list-trigger" },
      },
    },
    {
      id: "surface",
      role: "dialog",
      relativeTo: "list-trigger",
      resolve: {
        web: { kind: "class", className: ".poodle-popover__surface" },
        native: { kind: "id", id: "history-center:surface" },
      },
    },
    {
      // The bounded scroll region. Rows scroll inside it; the surface never
      // grows to the height of the history.
      id: "list",
      role: "list",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__list" },
        native: { kind: "id", id: "history-center:list" },
      },
    },
    {
      // Every visible entry row, spine or fork run. One part, one key space:
      // the row carries the hierarchy level, the fork identity travels on the
      // command payload, and nothing depends on indentation.
      id: "row",
      role: "listitem",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: '.poodle-history-center__row[data-row-kind="entry"]',
          keyAttribute: "data-row-entry",
        },
        native: { kind: "id-template", template: "history-center:row:{key}" },
      },
    },
    {
      id: "entry",
      role: "button",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: ".poodle-history-center__entry-content",
          keyAttribute: "data-entry",
        },
        native: { kind: "id-template", template: "history-center:entry:{key}" },
      },
    },
    {
      // The fork disclosure: a sibling of the entry button, never nested
      // inside it — no interactive element inside another interactive element.
      id: "disclosure",
      role: "button",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: ".poodle-history-center__fork",
          keyAttribute: "data-entry",
        },
        native: { kind: "id-template", template: "history-center:disclosure:{key}" },
      },
    },
    {
      // The picker row, keyed by the anchor whose forks it offers. Anchors are
      // entries, so it keys over the same sources.
      id: "picker",
      role: "listitem",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: '.poodle-history-center__row[data-row-kind="picker"]',
          keyAttribute: "data-row-entry",
        },
        native: { kind: "id-template", template: "history-center:picker:{key}" },
      },
    },
    {
      id: "picker-select",
      role: "combobox",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: ".poodle-select__trigger",
          keyAttribute: "data-anchor",
          keyScope: ".poodle-history-center__picker-controls",
        },
        native: { kind: "id-template", template: "history-center:picker-select:{key}" },
      },
    },
    {
      // One option per offered fork, keyed by the fork's first entry — the
      // continuation's stable identity, never a list index.
      id: "picker-option",
      role: "option",
      repeat: {
        sources: [{ hostRecord: "continuations" }],
        key: "entryId",
      },
      resolve: {
        web: {
          kind: "class",
          className: ".poodle-select__option",
          keyAttribute: "data-value",
        },
        native: { kind: "id-template", template: "history-center:picker-option:{key}" },
      },
    },
    {
      id: "picker-actions",
      role: "button",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: ".poodle-menu__trigger",
          keyAttribute: "data-anchor",
          keyScope: ".poodle-history-center__picker-actions",
        },
        native: { kind: "id-template", template: "history-center:picker-actions:{key}" },
      },
    },
    {
      // A singleton, not a repeated part: one actions menu is open at a
      // time, and `Menu` portals its surface out of the wrapper carrying the
      // anchor, so no ancestor rule could reach the items anyway.
      id: "action-rename",
      role: "menuitem",
      resolve: {
        web: {
          kind: "class",
          className: '.poodle-menu-surface__item[data-value="rename"]',
        },
        native: { kind: "id", id: "history-center:action-rename" },
      },
    },
    {
      // A singleton, not a repeated part: one actions menu is open at a
      // time, and `Menu` portals its surface out of the wrapper carrying the
      // anchor, so no ancestor rule could reach the items anyway.
      id: "action-checkout",
      role: "menuitem",
      resolve: {
        web: {
          kind: "class",
          className: '.poodle-menu-surface__item[data-value="checkout"]',
        },
        native: { kind: "id", id: "history-center:action-checkout" },
      },
    },
    {
      // The inline input that takes the Select's place while a rename is open.
      id: "rename-input",
      role: "textbox",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: ".poodle-history-center__rename-input",
          keyAttribute: "data-anchor",
        },
        native: { kind: "id-template", template: "history-center:rename-input:{key}" },
      },
    },
    {
      // An open fork whose run has not arrived. Never an empty gap, never a
      // dropped entry.
      id: "not-yet-loaded",
      role: "listitem",
      repeat: {
        sources: [{ prop: "pages", path: "entries" }, { hostRecord: "runEntries" }],
        key: "id",
      },
      resolve: {
        web: {
          kind: "class",
          className: '.poodle-history-center__row[data-row-kind="not-yet-loaded"]',
          keyAttribute: "data-row-entry",
        },
        native: { kind: "id-template", template: "history-center:not-yet-loaded:{key}" },
      },
    },
    {
      id: "empty",
      contains: "text",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__empty" },
        native: { kind: "id", id: "history-center:empty" },
      },
    },
    {
      id: "status",
      role: "status",
      contains: "text",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__loading" },
        native: { kind: "id", id: "history-center:status" },
      },
    },
    {
      id: "rejection",
      role: "status",
      contains: "text",
      resolve: {
        web: { kind: "class", className: ".poodle-history-center__rejection" },
        native: { kind: "id", id: "history-center:rejection" },
      },
    },
  ],
  states: [
    {
      name: "open",
      condition: "the surface is mounted and registered on the dismissable layer",
      web: "part-present",
      native: "part-present",
      part: "surface",
    },
    {
      name: "undoDisabled",
      condition: "undo is unavailable or an authority operation is running",
      web: "part-disabled-attr",
      native: "part-interaction-disabled",
      part: "undo",
    },
    {
      name: "redoDisabled",
      condition: "redo is unavailable or an authority operation is running",
      web: "part-disabled-attr",
      native: "part-interaction-disabled",
      part: "redo",
    },
    {
      name: "empty",
      condition: "no visible rows: pages absent or the derivation produced none",
      web: "part-present",
      native: "part-present",
      part: "empty",
    },
    {
      name: "rejected",
      condition: "a mapped rejection notice is displayed",
      web: "part-present",
      native: "part-present",
      part: "rejection",
    },
  ],
  tokenRoles: [
    { name: "placement", prop: "placement", default: "bottom-end" },
    { name: "status", prop: "status", default: "idle" },
    // HistoryCenter defaults to the chrome semantic role; at the default
    // presentation scale that resolves to sm in every active runtime.
    { name: "size", prop: "size", default: "sm" },
    { name: "density", prop: "density", default: "default" },
  ],
  axes: ["placement", "size", "density", "theme"],
  capabilities: [
    { name: "structure.identity", required: true },
    { name: "structure.part-resolution", required: true },
    { name: "overlay.intent", required: true },
    { name: "overlay.dismiss", required: true },
    { name: "overlay.layer", required: true },
    { name: "semantic.expanded", required: true },
    { name: "semantic.selected", required: true },
    { name: "semantic.disabled", required: true },
    { name: "semantic.token-roles", required: true },
    { name: "input.value", required: true },
    { name: "input.editing", required: true },
    { name: "focus", required: true },
    { name: "activate", required: true },
    { name: "interaction.key", required: true },
    { name: "layout.geometry", required: true },
    { name: "layout.position", required: true },
    { name: "accessibility.projection", required: true },
  ],
});

export type HistoryCenterInterface = typeof historyCenterInterface;
export type HistoryCenterPortableProps = PortablePropsOf<HistoryCenterInterface>;
export type HistoryCenterPortablePage =
  NonNullable<HistoryCenterPortableProps["pages"]>[number];
export type HistoryCenterPortableEntry = HistoryCenterPortablePage["entries"][number];
export type HistoryCenterPortableEvents = PortableEventsOf<HistoryCenterInterface>;
export type HistoryCenterPartId = PartIdsOf<HistoryCenterInterface>;
export type HistoryCenterEventName = EventNamesOf<HistoryCenterInterface>;
export type HistoryCenterTokenRole = TokenRoleNamesOf<HistoryCenterInterface>;
export type HistoryCenterAxis = AxisNamesOf<HistoryCenterInterface>;

export const HISTORY_CENTER_DEFAULT_PROPS: HistoryCenterPortableProps = {
  pages: null,
  canUndo: false,
  canRedo: false,
  busy: false,
  status: "idle",
  statusMessage: null,
  rejection: null,
  open: null,
  defaultOpen: false,
  placement: "bottom-end",
  undoLabel: "Undo",
  redoLabel: "Redo",
  listLabel: "History",
  title: "History",
  emptyMessage: "No history entries yet.",
  ariaLabel: null,
  maxBranchNameBytes: 256,
  size: null,
  sizeRole: "chrome",
  density: null,
};

File: packages/contracts/components/src/history_center.rs
//! HistoryCenter — the history counterpart to `MessageCenter`: a titlebar
//! trigger cluster (undo / list / redo) plus a popover rendering the flat
//! history list, with node-owned fork disclosure.
//!
//! This module is the single authority for the HistoryCenter declaration
//! surface: the struct, its defaults and builders, then the token recipes and
//! derived queries beside them. `g14.007` briefly generated the first half
//! from a TypeScript interface; `g14.008` rejected that path and `g14.021`
//! restored the hand-written declaration.
//!
//! The component is authority-agnostic. Data arrives through `pages`;
//! commands leave through the renderer's handlers. It validates no protocol
//! rule, assumes no supplied history is complete, and decides nothing about
//! what undo means. There is no Longhorn dependency and none is possible —
//! the dependency runs Longhorn → Poodle.
//!
//! Contract: `docs/contracts/components/history-center.md`

use crate::types::{ControlDensity, ControlSize};

#[derive(Clone, Copy, Debug, Eq, PartialEq, Default)]
pub enum HistoryCenterStatus {
    #[default] Idle,
    Loading,
    Failed,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq, Default)]
pub enum HistoryCenterRejection {
    #[default] AlreadyAtTarget,
    UnknownEntry,
    StaleHistory,
    ProtectedEntry,
    DeletionUnavailable,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct HistoryCenterSpec {
    pub pages: Option<Vec<crate::types::HistoryPathPage>>,
    pub can_undo: bool,
    pub can_redo: bool,
    pub is_busy: bool,
    pub status: HistoryCenterStatus,
    pub status_message: Option<String>,
    pub rejection: Option<HistoryCenterRejection>,
    pub open: Option<bool>,
    pub default_open: bool,
    pub placement: crate::types::OverlayPlacement,
    pub undo_label: String,
    pub redo_label: String,
    pub list_label: String,
    pub title: String,
    pub empty_message: String,
    pub aria_label: Option<String>,
    pub max_branch_name_bytes: usize,
    pub size: Option<crate::types::ControlSize>,
    pub size_role: crate::types::SemanticControlSizeRole,
    pub density: Option<crate::types::ControlDensity>,
}

impl Default for HistoryCenterSpec {
    fn default() -> Self {
        Self {
            pages: None,
            can_undo: false,
            can_redo: false,
            is_busy: false,
            status: HistoryCenterStatus::Idle,
            status_message: None,
            rejection: None,
            open: None,
            default_open: false,
            placement: crate::types::OverlayPlacement::BottomEnd,
            undo_label: "Undo".to_owned(),
            redo_label: "Redo".to_owned(),
            list_label: "History".to_owned(),
            title: "History".to_owned(),
            empty_message: "No history entries yet.".to_owned(),
            aria_label: None,
            max_branch_name_bytes: 256,
            size: None,
            size_role: crate::types::SemanticControlSizeRole::Chrome,
            density: None,
        }
    }
}

impl HistoryCenterSpec {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_pages(mut self, value: Vec<crate::types::HistoryPathPage>) -> Self {
        self.pages = Some(value);
        self
    }
    pub fn with_can_undo(mut self, value: bool) -> Self {
        self.can_undo = value;
        self
    }
    pub fn with_can_redo(mut self, value: bool) -> Self {
        self.can_redo = value;
        self
    }
    pub fn with_busy(mut self, value: bool) -> Self {
        self.is_busy = value;
        self
    }
    pub fn with_status(mut self, value: HistoryCenterStatus) -> Self {
        self.status = value;
        self
    }
    pub fn with_status_message(mut self, value: impl Into<String>) -> Self {
        self.status_message = Some(value.into());
        self
    }
    pub fn with_rejection(mut self, value: HistoryCenterRejection) -> Self {
        self.rejection = Some(value);
        self
    }
    pub fn with_open(mut self, value: bool) -> Self {
        self.open = Some(value);
        self
    }
    pub fn with_default_open(mut self, value: bool) -> Self {
        self.default_open = value;
        self
    }
    pub fn with_placement(mut self, value: crate::types::OverlayPlacement) -> Self {
        self.placement = value;
        self
    }
    pub fn with_undo_label(mut self, value: impl Into<String>) -> Self {
        self.undo_label = value.into();
        self
    }
    pub fn with_redo_label(mut self, value: impl Into<String>) -> Self {
        self.redo_label = value.into();
        self
    }
    pub fn with_list_label(mut self, value: impl Into<String>) -> Self {
        self.list_label = value.into();
        self
    }
    pub fn with_title(mut self, value: impl Into<String>) -> Self {
        self.title = value.into();
        self
    }
    pub fn with_empty_message(mut self, value: impl Into<String>) -> Self {
        self.empty_message = value.into();
        self
    }
    pub fn with_aria_label(mut self, value: impl Into<String>) -> Self {
        self.aria_label = Some(value.into());
        self
    }
    pub fn with_max_branch_name_bytes(mut self, value: usize) -> Self {
        self.max_branch_name_bytes = value;
        self
    }
    pub fn with_size(mut self, value: crate::types::ControlSize) -> Self {
        self.size = Some(value);
        self
    }
    pub fn with_size_role(mut self, value: crate::types::SemanticControlSizeRole) -> Self {
        self.size_role = value;
        self
    }
    pub fn with_density(mut self, value: crate::types::ControlDensity) -> Self {
        self.density = Some(value);
        self
    }
}


impl HistoryCenterSpec {
    /// The resolved control size: the explicit size when set, else the
    /// inherited presentation scale stepped by the semantic role.
    pub fn resolved_size(&self, inherited: ControlSize) -> ControlSize {
        match self.size {
            Some(size) => size,
            None => crate::types::resolve_semantic_control_size(inherited, self.size_role),
        }
    }

    pub fn resolved_density(&self, inherited: ControlDensity) -> ControlDensity {
        self.density.unwrap_or(inherited)
    }

    /// The surface's accessible name: the override when supplied, else the
    /// heading. A surface without either would announce as an unnamed dialog.
    pub fn surface_label(&self) -> &str {
        self.aria_label.as_deref().unwrap_or(&self.title)
    }

    /// Whether the undo trigger is inert: the host says undo is unavailable,
    /// or an authority operation is already running.
    pub fn undo_is_disabled(&self) -> bool {
        !self.can_undo || self.is_busy
    }

    pub fn redo_is_disabled(&self) -> bool {
        !self.can_redo || self.is_busy
    }

    /// The display copy for a mapped rejection. The host's bridge maps its
    /// protocol onto the five codes; the component owns the wording, so the
    /// protocol's vocabulary never reaches an operator. The three deletion
    /// refusals stay distinct from `UnknownEntry`: a stale, protected, or
    /// unavailable deletion is not a missing entry.
    pub fn rejection_message(&self) -> Option<&'static str> {
        match self.rejection? {
            HistoryCenterRejection::AlreadyAtTarget => Some("Already at the requested target"),
            HistoryCenterRejection::UnknownEntry => Some("Entry does not exist"),
            HistoryCenterRejection::StaleHistory => {
                Some("History changed; this entry was not deleted")
            }
            HistoryCenterRejection::ProtectedEntry => Some("This history entry is protected"),
            HistoryCenterRejection::DeletionUnavailable => Some("History deletion is unavailable"),
        }
    }

    /// The status row's copy: the host's message when the source failed, else
    /// the component's own loading line.
    pub fn status_line(&self) -> Option<&str> {
        match self.status {
            HistoryCenterStatus::Idle => None,
            HistoryCenterStatus::Loading => {
                Some(self.status_message.as_deref().unwrap_or("Loading history…"))
            }
            HistoryCenterStatus::Failed => Some(
                self.status_message
                    .as_deref()
                    .unwrap_or("History unavailable"),
            ),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn triggers_are_inert_without_availability_or_while_busy() {
        let spec = HistoryCenterSpec::new();
        assert!(spec.undo_is_disabled());
        assert!(spec.redo_is_disabled());

        let ready = HistoryCenterSpec::new().with_can_undo(true).with_can_redo(true);
        assert!(!ready.undo_is_disabled());
        assert!(!ready.redo_is_disabled());

        let busy = ready.clone().with_busy(true);
        assert!(busy.undo_is_disabled());
        assert!(busy.redo_is_disabled());
    }

    #[test]
    fn the_surface_falls_back_to_its_heading_for_a_name() {
        let spec = HistoryCenterSpec::new();
        assert_eq!(spec.surface_label(), "History");
        assert_eq!(
            spec.with_aria_label("Project history").surface_label(),
            "Project history",
        );
    }

    /// The exact contract table (`docs/contracts/components/history-center.md`
    /// §"Rejection handling"). Both rejection proofs below read this list, so
    /// dropping a category or pointing two codes at one message fails here.
    const REJECTION_COPY: [(HistoryCenterRejection, &str); 5] = [
        (
            HistoryCenterRejection::AlreadyAtTarget,
            "Already at the requested target",
        ),
        (HistoryCenterRejection::UnknownEntry, "Entry does not exist"),
        (
            HistoryCenterRejection::StaleHistory,
            "History changed; this entry was not deleted",
        ),
        (
            HistoryCenterRejection::ProtectedEntry,
            "This history entry is protected",
        ),
        (
            HistoryCenterRejection::DeletionUnavailable,
            "History deletion is unavailable",
        ),
    ];

    #[test]
    fn rejection_copy_is_component_owned() {
        assert_eq!(HistoryCenterSpec::new().rejection_message(), None);
        for (code, message) in REJECTION_COPY {
            assert_eq!(
                HistoryCenterSpec::new()
                    .with_rejection(code)
                    .rejection_message(),
                Some(message),
                "{code:?} must carry its own component-owned copy",
            );
        }
    }

    /// The papercut this replaced: a stale revision, a protected entry, and an
    /// unavailable deletion all told the operator "Entry does not exist".
    #[test]
    fn every_refusal_meaning_stays_distinct() {
        let mut messages: Vec<&str> = REJECTION_COPY.iter().map(|(_, m)| *m).collect();
        let distinct = messages.len();
        messages.sort_unstable();
        messages.dedup();
        assert_eq!(messages.len(), distinct, "no two codes may share copy");

        let unknown = HistoryCenterSpec::new()
            .with_rejection(HistoryCenterRejection::UnknownEntry)
            .rejection_message();
        for code in [
            HistoryCenterRejection::StaleHistory,
            HistoryCenterRejection::ProtectedEntry,
            HistoryCenterRejection::DeletionUnavailable,
        ] {
            assert_ne!(
                HistoryCenterSpec::new().with_rejection(code).rejection_message(),
                unknown,
                "{code:?} is a refused deletion, not a missing entry",
            );
        }
    }

    #[test]
    fn a_failed_source_prefers_the_hosts_message() {
        let spec = HistoryCenterSpec::new().with_status(HistoryCenterStatus::Failed);
        assert_eq!(spec.status_line(), Some("History unavailable"));
        assert_eq!(
            spec.with_status_message("Source offline").status_line(),
            Some("Source offline"),
        );
    }

    #[test]
    fn chrome_steps_the_inherited_scale_down() {
        let spec = HistoryCenterSpec::new();
        assert_eq!(spec.resolved_size(ControlSize::Md), ControlSize::Sm);
        assert_eq!(
            spec.with_size(ControlSize::Lg).resolved_size(ControlSize::Md),
            ControlSize::Lg,
        );
    }
}
